src/components/app/app.ts

```diff
@@ -0,0 +1,102 @@
+import '../models/StoreModel';
+import StoreModel from '../models/StoreModel';
+import Footer from '../views/FooterVew';
+import CartModel from '../models/CartModel';
+
+export default class AppController implements AppControllerInterface {
+    private static instance: InstanceType<typeof AppController>;
+    public mainContainer: HTMLElement;
+    public header: Header;
+    public routes: Route[];
+    public router: Router;
+    public cartView: CartView;
+    public storeView: StoreView;
+    public startPage: StartPageView;
+    public menu: Menu;
+    public productPage: ProductPage;
+    public storeModel: StoreModel;
+    public footer: Footer;
+    public cartModel: CartModel;
+
+    constructor() {
+        this.menu = new Menu(this);
+        this.header = new Header(this);
+        this.cartModel = new CartModel(this);
+        this.cartView = new CartView(this.cartModel, this);
+        this.mainContainer = createElem('main', 'main wrapper');
+        this.startPage = new StartPageView(this);
+        this.storeModel = new StoreModel(this);
+        this.storeView = new StoreView(this.storeModel, this);
+        this.productPage = new ProductPage(this);
+        this.footer = new Footer();
+        this.routes = [
+            new Route('', LINKS.About, this.startPage),
+            new Route('store', LINKS.Store, this.storeView),
+            new Route('cart', LINKS.Cart, this.cartView),
+            new Route('product', '#product', this.productPage),
+        ];
+        this.router = new Router(this, this.routes);
+
+        if (AppController.instance) {
+            return AppController.instance;
+        }
+        AppController.instance = this;
+    }
+
+    public start() {
+        document.body.append(this.header.createHeader(), this.mainContainer, this.footer.renderFooter());
+        this.router.init();
+        this.addToLocalStorage();
+    }
+
+    public updatePage(view: ViewComponent, params?: Props) {
+        this.destroyAllChildNodes(this.mainContainer);
+        this.mainContainer.append(view.render(params));
+        this.header.setActiveLink();
+    }
+
+    public addProductToCart(product: Product) {
+        this.cartModel.addProduct(product);
+        this.cartView.updateCartInfo();
+        this.cartView.updatePromoBlock();
+    }
+
+    public deleteProductFromCart(product: Product) {
+        this.cartModel.deleteProduct(product);
+        this.cartView.updateCartInfo();
+        this.cartView.updatePromoBlock();
+    }
+
+    private addToLocalStorage() {
+        window.addEventListener('beforeunload', () => {
+            localStorage.setItem('cart', JSON.stringify(this.cartModel.productsInCart));
+            localStorage.setItem('totalSum', this.cartModel.totalSum.toString());
+            localStorage.setItem('productsQuantity', this.cartModel.productsQuantity.toString());
+            localStorage.setItem('activatedPromocodes', JSON.stringify(this.cartModel.activatedPromocodes));
```

Comment:
Wouldn't it be better tu use one object with all these properties?

Reply:
refactored local storage methods (added common state object, typed it as interface and deleted excess code)